source/encoding.c

```diff
@@ -216,73 +227,104 @@ int aws_hex_decode(const struct aws_byte_cursor *AWS_RESTRICT to_decode, struct
      * Rounding is because the output will be padded with '=' chars if necessary to make it divisible by 4. */
 
     /* adding 2 before dividing by 3 is a trick to round up during division */
-    size_t tmp = to_encode_len + 2;
+    size_t tmp = 0;
 
-    if (AWS_UNLIKELY(tmp < to_encode_len)) {
-        return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+    if (AWS_UNLIKELY(aws_add_size_checked(to_encode_len, 2, &tmp))) {
+        return AWS_OP_ERR;
     }
 
     tmp /= 3;
-    size_t overflow_check = tmp;
-    tmp = 4 * tmp;
+    if (AWS_UNLIKELY(aws_mul_size_checked(tmp, 4, &tmp))) {
+        return AWS_OP_ERR;
+    }
 
-    if (AWS_UNLIKELY(tmp < overflow_check)) {
-        return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+    *encoded_len = tmp;
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_base64_url_compute_encoded_len(size_t to_encode_len, size_t *encoded_len) {
+    AWS_ERROR_PRECONDITION(encoded_len);
+
+    /* Just do direct math for each each 6 bits map to char. +5 is same trick to as before to round up. */
+
+    size_t tmp = 0;
+
+    if (AWS_UNLIKELY(aws_mul_size_checked(to_encode_len, 8, &tmp))) {
+        return AWS_OP_ERR;
+    }
+
+    if (AWS_UNLIKELY(aws_add_size_checked(tmp, 5, &tmp))) {
+        return AWS_OP_ERR;
     }
 
+    tmp /= 6;
+
     *encoded_len = tmp;
 
     return AWS_OP_SUCCESS;
 }
 
+bool s_ispadding(uint8_t ch) {
+    return ch == '=';
+}
+
 int aws_base64_compute_decoded_len(const struct aws_byte_cursor *AWS_RESTRICT to_decode, size_t *decoded_len) {
     AWS_ASSERT(to_decode);
     AWS_ASSERT(decoded_len);
 
-    const size_t len = to_decode->len;
-    const uint8_t *input = to_decode->ptr;
+    struct aws_byte_cursor trimmed = aws_byte_cursor_right_trim_pred(to_decode, s_ispadding);
+
+    const size_t len = trimmed.len;
 
     if (len == 0) {
         *decoded_len = 0;
         return AWS_OP_SUCCESS;
     }
 
-    /* ensure it's divisible by 4 */
-    if (AWS_UNLIKELY(len & 0x03)) {
+    /* impossible len */
+    if (AWS_UNLIKELY(len % 4 == 1)) {
         return aws_raise_error(AWS_ERROR_INVALID_BASE64_STR);
     }
 
     /* For every 4 ascii characters of encoded input, there will be 3 bytes of decoded output (deal with padding later)
      * decoded_len = 3/4 * len       <-- note that result will be smaller then len, so overflow can be avoided
      *             = (len / 4) * 3   <-- divide before multiply to avoid overflow
      */
-    size_t decoded_len_tmp = (len / 4) * 3;
+    size_t decoded_len_tmp = 0;
 
-    /* But last two ascii chars might be padding. */
-    AWS_ASSERT(len >= 4); /* we checked earlier len != 0, and was divisible by 4 */
-    size_t padding = 0;
-    if (input[len - 1] == '=' && input[len - 2] == '=') { /*last two chars are = */
-        padding = 2;
-    } else if (input[len - 1] == '=') { /*last char is = */
-        padding = 1;
+    if (aws_mul_size_checked(len, 3, &decoded_len_tmp)) {
```

Comment:
The above comment mentioned we should divide by 4 first to avoid overflow, but we are doing multiply first here?